specs/logging_implementation.adoc

```diff
@@ -0,0 +1,273 @@
+* A schema needs to be defined for emitted messages.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must pass the message as per schema.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+* Store the operation details in etcd:
+```
+1. messages with inorder_id
+    {job_id}/{in_order_id}
+```
+
+== Impacted modules
+
+=== Tendrl/common impact
+* Create a new class called "Message" in common for message handling with attributes
+  handler_name, message_type, message, log_level, timestamp(format: ISO 8601).
+* Create a new class called "Log" in common and inherit it from Message class.
+* Create a new class called "Alert" in common with attributes alert_id, node_id,
+  resource, current_value, tags, type, severity, significance, ackedby, acked.
+  Inherit alert class from Message class.
+* Create a new class called "Operation" in common with attributes
+  job_id, operation_summary. Inherit Operation class from Message class.
+```
+Alert class object structure
+  - alert_id
+  - node_id
+  - resource
+  - current_value
+  - tags
+  - type
+  - severity
+  - significance
+  - ackedby
+  - acked
+
+Operation class object structure
+  - log_handler_name
+  - job_id
+  - log_level
+  - operation_summary
+
+Log class object structure
+  - log_handler_name
+  - log_level
+  - log_message
+
+Message class object structure
+  - message_type
+  - timestamp (format : ISO 8601)
+```
+* Create a new function called message_handler in Message class and process the
+  object values.
+* If the message type is Alert then put the object in node_agent socket, logging
+  for alert is handled from alert handler.
+* If the message type is operation then log the message based on log level with
+  message job id (job id is used to trace the operation) and store the object in etcd.
+* If the type is Log then store the message in log file based on log level.
+
+=== Tendrl/node_agent impact
```

Comment:
@nnDarshan It's because other processes would want to be able to push their messages through the node agent. Anyway, @anmolbabu has already worked on the alerts socket, the same can be resued.